lib/ingress/ingress.ts

```diff
@@ -0,0 +1,143 @@
+    if (supportsTls(props)) {
+      if (props.tlsDomain) {
+        ingressTls.push({
+          hosts: ensureArray(props.tlsDomain),
+        });
+      }
+      ingressListenPorts.push({ HTTPS: 443 });
+    }
+
+    if (props.externalHostname) {
+      externalDns["external-dns.alpha.kubernetes.io/hostname"] =
+        props.externalHostname;
+    }
+
+    const targetGroups: Record<string, string | number>[] = [];
+    for (const service of props.serviceRouting) {
+      const serviceName = `${service.name}-${service.namespace}`;
+      new KubeService(this, `${id}-${serviceName}-service`, {
+        metadata: {
+          name: serviceName,
+          namespace: "application-ingress",
+        },
+        spec: {
+          type: ServiceSpecType.EXTERNAL_NAME,
+          externalName: `${service.name}.${service.namespace}.svc.cluster.local`,
+          ports: [
+            {
+              port: service.port,
+              targetPort: IntOrString.fromNumber(service.port),
+            },
+          ],
+        },
+      });
+
+      const targetGroup = {
+        serviceName: serviceName,
+        servicePort: service.port,
+        weight: service.weight,
+      };
+      targetGroups.push(targetGroup);
+    }
+
+    const ingressAnnotations = {
+      "alb.ingress.kubernetes.io/listen-ports":
+        convertToJsonContent(ingressListenPorts),
+      "alb.ingress.kubernetes.io/success-codes": "200,303",
+      "alb.ingress.kubernetes.io/target-type": ingressTargetType,
+      "alb.ingress.kubernetes.io/tags": convertToStringMap({
+        service: labels.service ?? app,
+        instance: id,
+        environment: environment,
+      }),
+      "alb.ingress.kubernetes.io/actions.service-weighting":
+        convertToJsonContent({
+          type: "forward",
+          forwardConfig: {
+            targetGroups: targetGroups,
+          },
+        }),
+      ...ingressTlsAnnotations,
+      ...props.ingressAnnotations,
+      ...externalDns,
+    };
+
+    for (const hostname of props.hostnames) {
+      ingressRules.push({
+        host: hostname,
+        http: {
+          paths: [
+            {
+              pathType: "Prefix",
+              path: "/",
+              backend: {
+                service: {
+                  name: "service-weighting",
+                  port: {
+                    name: "use-annotation",
+                  },
+                },
+              },
+            },
+          ],
+        },
+      });
+    }
+
+    new KubeIngress(this, `${id}-ingress`, {
+      metadata: {
+        annotations: ingressAnnotations,
+        labels: labels,
+      },
+      spec: {
+        ingressClassName: props.ingressClassName,
+        tls: ingressTls.length > 0 ? ingressTls : undefined,
+        rules: ingressRules.length > 0 ? ingressRules : undefined,
+      },
+    });
+  }
+}
+
+function convertToJsonContent(obj: unknown): string {
+  return JSON.stringify(obj);
+}
```

Comment:
You can import the helpers already present in the library:
https://github.com/talis/talis-cdk8s-constructs/blob/526214c97af149ef673a3e957da4220d59f0fb76/lib/web-service/annotation-util.ts#L85-L92

You may move `annotation-util.ts` into `lib/common/` directory.